services/data-info/src/data_info/services/permissions.clj

```diff
@@ -16,16 +17,52 @@
      #(not (contains? filtered-users (:user %1)))
      (perm/list-user-perm cm abspath))))
 
+(defn list-permissions*
+  [cm user path]
+  {:user-permissions (filtered-user-perms cm user path)})
+
 (defn list-permissions
   [{:keys [user]} data-id]
   (with-jargon (cfg/jargon-cfg) [cm]
     (let [path (ft/rm-last-slash (:path (uuids/path-for-uuid user data-id)))]
       (validators/user-exists cm user)
       (validators/path-readable cm user path)
-      {:user-permissions (filtered-user-perms cm user path)})))
+      (list-permissions* cm user path))))
 
 (with-pre-hook! #'list-permissions
   (fn [params data-id]
     (dul/log-call "list-permissions" params data-id)))
 
 (with-post-hook! #'list-permissions (dul/log-func "list-permissions"))
+
+(defn add-permission
+  [{:keys [user]} data-id share-with permission]
+  (with-jargon (cfg/jargon-cfg) [cm]
+    (let [path (ft/rm-last-slash (:path (uuids/path-for-uuid user data-id)))]
+      (validators/user-exists cm user)
+      (validators/user-owns-path cm user path)
+      (validators/user-exists cm share-with)
```

Comment:
I think I don't see where an `ERR_NOT_READABLE` can happen here (which makes some sense, since there's not much reason to check ownership and readability both). Can probably be removed from the swagger documentation (same for remove-permission).